File: Data Structures & Algorithms/binary-tree-diameter/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_diameter = 0
        
        def get_height(node):
            if not node:
                return 0
            
            # 1. Ask left and right subtrees for their heights
            left_height = get_height(node.left)
            right_height = get_height(node.right)
            
            # 2. Update the global maximum diameter found so far
            self.max_diameter = max(self.max_diameter, left_height + right_height)
            
            # 3. Return the height of the current node to its parent
            return 1 + max(left_height, right_height)
        
        get_height(root)
        return self.max_diameter

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int dfs(TreeNode * root, int max){
        if(!root) return 0;
        if(root->val >= max){
            return 1+dfs(root->right, root->val)+dfs(root->left, root->val);
        }
            return dfs(root->right, max)+dfs(root->left,max);
    }
    int goodNodes(TreeNode* root) {
        //top down prob 
        int r = dfs(root->right, root->val);
        int l = dfs(root->left, root->val);
        return r+l+1;

    }
};
